modules/core/04-channel/keeper/upgrade.go

```diff
@@ -90,3 +99,41 @@ func (k Keeper) WriteUpgradeInitChannel(
 
 	emitChannelUpgradeInitEvent(ctx, portID, channelID, upgradeSequence, channelUpgrade)
 }
+
+// RestoreChannel restores the given channel to the state prior to upgrade.
+func (k Keeper) RestoreChannel(ctx sdk.Context, portID, channelID string, upgradeSequence uint64) error {
+	_, code, _ := errorsmod.ABCIInfo(types.ErrChannelRestored, false) // discard non-determinstic codespace and log values
+
+	errorReceipt := types.ErrorReceipt{
+		Sequence: upgradeSequence,
+		Error:    fmt.Sprintf("ABCI code: %d: %s", code, restoreErrorString),
+	}
+
+	k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+
+	channel, found := k.GetUpgradeRestoreChannel(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrRestoreChannelNotFound, "channel-id: %s", channelID)
+	}
+
+	k.SetChannel(ctx, portID, channelID, channel)
+	k.DeleteUpgradeRestoreChannel(ctx, portID, channelID)
+	k.DeleteUpgradeTimeout(ctx, portID, channelID)
+
+	module, _, err := k.LookupModuleByChannel(ctx, portID, channelID)
+	if err != nil {
+		return errorsmod.Wrap(err, "could not retrieve module from port-id")
+	}
+
+	portKeeper, ok := k.portKeeper.(*portkeeper.Keeper)
+	if !ok {
+		panic("todo: handle this situation")
+	}
+
+	cbs, found := portKeeper.Router.GetRoute(module)
+	if !found {
+		return errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module)
+	}
+
+	return cbs.OnChanUpgradeRestore(ctx, portID, channelID)
```

Comment:
I know this is the logic that the spec specifies at the moment, but I am wondering if we could change the implementation (maybe the spec as well) to do the application callback in the `ChannelUpgradeTry` function in `core/keeper/msg_server.go` so that either we execute there the callback for success or the callback for restore depending on the error returned by `ChanUpgradeTry` in `04-channel/keeper/upgrade.go`. I think that would make the code cleaner, since we can keep it consistent at which layer we execute application callbacks.